compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1479,9 +1479,39 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
 
     /** Like tp1 <:< tp2, but returns false immediately if we know that
      *  the case was covered previously during subtyping.
+     *
+     *  A type has been covered previously in subtype checking if it
+     *  is some combination of TypeRefs that point to classes, where the
+     *  combiners are AppliedTypes, RefinedTypes, RecTypes, And/Or-Types or AnnotatedTypes.
+     *
+     *  The exception is that if both sides contain OrTypes, the check hasn't been covered.
+     *  See #17465.
      */
     def isNewSubType(tp1: Type): Boolean =
-      if (isCovered(tp1) && isCovered(tp2))
+
+      def isCovered(tp: Type): (Boolean, Boolean) =
+        var containsOr: Boolean = false
+        @annotation.tailrec def recur(todos: List[Type]): Boolean = todos match
+          case tp :: todos =>
+            tp.dealiasKeepRefiningAnnots.stripTypeVar match
+              case tp: TypeRef =>
+                if tp.symbol.isClass && tp.symbol != NothingClass && tp.symbol != NullClass then recur(todos)
+                else false
+              case tp: AppliedType => recur(tp.tycon :: todos)
+              case tp: RefinedOrRecType => recur(tp.parent :: todos)
+              case tp: AndType => recur(tp.tp1 :: tp.tp2 :: todos)
+              case tp: OrType  =>
+                containsOr = true
+                recur(tp.tp1 :: tp.tp2 :: todos)
+              case _ => false
+          case Nil => true
+        val result = recur(tp :: Nil)
+        (result, containsOr)
+
+      val (covered1, hasOr1) = isCovered(tp1)
+      val (covered2, hasOr2) = isCovered(tp2)
+
+      if covered1 && covered2 && !(hasOr1 && hasOr2) then
```

Comment:
So total, 4 allocations. We should avoid them in subtype checks, where possible. 

Techniques do so:

 - Instead of a variable, pass a second parameter to `recur`.
 - Use an enum instead of a pair of booleans. We need three states: `Uncovered, CoveredWithOr, Covered`